### PR TITLE
Fix harbinger file deletion race condition (resolves #1046)

### LINKED ISSUE #1046: Harbingerfile deletion race condition
```
l/k/jobIi7VIc:  Traceback (most recent call last):
l/k/jobIi7VIc:    File "/home/alden/progressiveCactus3/python/local/lib/python2.7/site-packages/toil/worker.py", line 332, in main
l/k/jobIi7VIc:      job._runner(jobWrapper=jobWrapper, jobStore=jobStore, fileStore=fileStore)
l/k/jobIi7VIc:    File "/home/alden/progressiveCactus3/python/local/lib/python2.7/site-packages/toil/job.py", line 2323, in _runner
l/k/jobIi7VIc:      returnValues = self._run(jobWrapper, fileStore)
l/k/jobIi7VIc:    File "/home/alden/progressiveCactus3/python/local/lib/python2.7/site-packages/toil/job.py", line 2271, in _run
l/k/jobIi7VIc:      return self.run(fileStore)
l/k/jobIi7VIc:    File "/home/alden/progressiveCactus3/submodules/cactus/pipeline/cactus_workflow.py", line 900, in run
l/k/jobIi7VIc:      self.cactusSequencesPath = fileStore.readGlobalFile(self.cactusSequencesID)
l/k/jobIi7VIc:    File "/home/alden/progressiveCactus3/python/local/lib/python2.7/site-packages/toil/job.py", line 816, in readGlobalFile
l/k/jobIi7VIc:      harbingerFile.waitOnDownload(lockFileHandle)
l/k/jobIi7VIc:    File "/home/alden/progressiveCactus3/python/local/lib/python2.7/site-packages/toil/job.py", line 1603, in waitOnDownload
l/k/jobIi7VIc:      pid = self.read()
l/k/jobIi7VIc:    File "/home/alden/progressiveCactus3/python/local/lib/python2.7/site-packages/toil/job.py", line 1616, in read
l/k/jobIi7VIc:      return int(open(self.harbingerFileName).read())
l/k/jobIi7VIc:  IOError: [Errno 2] No such file or directory: '/tmp/toil-269fb642-c572-42a1-8c9f-aa751fa3e5e9/cache-269fb642-c572-42a1-8c9f-aa751fa3e5e9/.di9QL3RtcHo0d3RtVC50bXA=.harbinger'
```

I haven't fully investigated this yet, but it looks like the harbinger file can be deleted by the asyncWrite thread after self.exists() is checked but before self.read() is called in waitOnDownload(). Maybe HarbingerFile.delete() should acquire the cache lock? Also, should line https://github.com/BD2KGenomics/toil/blob/edf86b5e253e9003a441a68773e79ecaded1a599/src/toil/job.py#L1613 be self.delete() instead of harbingerFile.delete()?

### PR DESCRIPTION
**Pull request by arkal**
resolves #1046

Harbinger file deletion now occurs only with the cache lock held.

### CODE REVIEW

**Comment by arkal** [comment]
@adderan  can you confirm this works?

**Comment by adderan** [comment]
It works. Thanks!

File: src/toil/job.py [hunk 1]
```diff
@@ -1578,6 +1577,7 @@ def __init__(self, fileStore, fileStoreID=None, cachedFileName=None):
                     cachedFileName = fileStore.encodedFileID(fileStoreID)
                 else:
                     self.fileStoreID = fileStore.decodedFileID(cachedFileName)
+                self.fileStore = fileStore
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Just noticed that the docstring incorrectly declares a return value.

File: src/toil/job.py [hunk 2]
```diff
@@ -1610,7 +1610,9 @@ def waitOnDownload(self, lockFileHandle):
                     else:
                         # The process that was supposed to download the file has died so we need
                         # to remove the harbinger.
-                        harbingerFile.delete()
+                        # Release the lock on the cache lock file first
```
Comment (hannes-ucsc) [inline src/toil/job.py]: This needs an explanation, I think.

Comment (hannes-ucsc) [inline src/toil/job.py]: Never mind.

**Comment by arkal** [comment]
@hannes-ucsc i implemented the suggested changes. This is ready for a review

File: src/toil/job.py [hunk 3]
```diff
@@ -852,8 +852,7 @@ def readGlobalFile(self, fileStoreID, userPath=None, cache=True, mutable=None):
                                 # already does it for us
                         finally:
                             # Reacquire the file lock and delete the partial file.
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Comment is now inaccurate.

**Comment by hannes-ucsc** [comment]
We must have misunderstood each other. This is way too much code for the "quick and easy" solution. Making the lock reentrant will be considerably less code and less complicated. Let me know if you want me to do that refactoring.

File: src/toil/job.py (hunk 3 above)
Comment (arkal) [inline src/toil/job.py]: but harbingerFile.delete does reacquire the lock and delete the file.

**Comment by arkal** [comment]
But this is almost exactly what we discussed. We had talked about `deleteWithLock` and `delete` where `deleteWithLock` would call `delete`. Instead I made is `delete` and `_delete`. All the other code added is just debug logs that weren't there before that I felt was useful.

File: src/toil/job.py (hunk 3 above)
Comment (hannes-ucsc) [inline src/toil/job.py]: That's an implementation detail that shouldn't be repeated here.

**Comment by hannes-ucsc** [comment]
Not true. You added PID checking, which adds complexity and the need for logging that otherwise adds no value.

If you never release the lock after determining that the harbinger file needs to be acted upon, none of those races can occur. If a quick-and-dirty fix doesn't guarantee that, and instead requires PID checking—which I doubt, BTW—then we need to implement reentrant locking as a clean solution.

**Comment by arkal** [comment]
I do remember talking about the  pid checking.... the edge case where two readers reach the deletion of the harbinger at the same time..

**Comment by arkal** [comment]
OK on nth thought, we don't need PID checking if we strictly check the pids before deleting  only with a lock held (which is enforced in `waitOnDownload`). The other 2 delete conditions don't need to check pids because the deleting thread/process is the one that created the file (asyncWrite and readGlobalFile).

File: src/toil/job.py (hunk 3 above)
Comment (arkal) [inline src/toil/job.py]: okai :)

**Comment by arkal** [comment]
Jenkins, test this please.

**Comment by arkal** [comment]
@hannes-ucsc lines like 836 and 859 will become problematic with reentrant locks. I'm trying to stay away from that till a later sprint. I think we might be able to agree on the current PR. It's (hopefully) exactly what we discussed this morning.

**Comment by arkal** [comment]
@hannes-ucsc this is ready!

**Comment by hannes-ucsc** [comment]
@arkal, I suggest that you convert the line references in [your comment](https://github.com/BD2KGenomics/toil/pull/1047#issuecomment-231314990) into clickable links, referring to a fixed commit. That way they won't get stale in, like, 2 seconds.

**Comment by arkal** [comment]
done

**Pull request merged**